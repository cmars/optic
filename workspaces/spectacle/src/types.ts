import { ExecutionResult } from 'graphql';
import { EventEmitter } from 'events';

import {
  ILearnedBodies,
  IAffordanceTrailsDiffHashMap,
} from '@useoptic/cli-shared/build/diffs/initial-types';
import {
  IHttpInteraction,
  ShapeChoice,
  JsonType,
} from '@useoptic/optic-domain';
import { endpoints, shapes } from '@useoptic/graph-lib';
import { CommandGenerator, ContributionsProjection } from './helpers';
import { IOpticCommandContext } from './in-memory';

////////////////////////////////////////////////////////////////////////////////

export interface IOpticEngine {
  try_apply_commands(
    commandsJson: string,
    eventsJson: string,
    batchId: string,
    commitMessage: string,
    clientId: string,
    clientSessionId: string
  ): any;

  affordances_to_commands(
    json_affordances_json: string,
    json_trail_json: string,
    id_generator_strategy: string
  ): string;

  get_shape_viewer_projection(spec: any): string;

  get_contributions_projection(spec: any): string;

  learn_shape_diff_affordances(
    spec: any,
    diff_results_json: string,
    tagged_interactions_jsonl: string
  ): string;

  learn_undocumented_bodies(
    spec: any,
    interactions_jsonl: string,
    id_generator_strategy: string
  ): string;

  spec_from_events(eventsJson: string): any;

  spec_endpoint_delete_commands(
    spec: any,
    path_id: string,
    method: string
  ): string;

  spec_field_remove_commands(spec: any, field_id: string): string;

  spec_field_edit_commands(
    spec: any,
    field_id: string,
    requestedTypes: JsonType[],
    id_generator_strategy: IOpticEngineIdGenerationStrategy
  ): string;
}

export enum IOpticEngineIdGenerationStrategy {
  SEQUENTIAL = 'sequential',
  UNIQUE = 'unique',
}

////////////////////////////////////////////////////////////////////////////////

export interface IOpticSpecRepository {
  listEvents(): Promise<any[]>;
}

export interface IOpticSpecReadWriteRepository extends IOpticSpecRepository {
  applyCommands(
    commands: any[],
    batchCommitId: string,
    commitMessage: string,
    commandContext: IOpticCommandContext
  ): Promise<void>;

  changes: AsyncGenerator<number>;

  notifications: EventEmitter;

  resetToCommit(batchCommitId: string): Promise<void>;
}

////////////////////////////////////////////////////////////////////////////////

export interface ICapture {
  captureId: string;
  startedAt: string;
}

export interface StartDiffResult {
  //notifications: EventEmitter;
  onComplete: Promise<IOpticDiffService>;
}

export interface IOpticCapturesService {
  listCaptures(): Promise<ICapture[]>;

  loadInteraction(
    captureId: string,
    pointer: string
  ): Promise<IHttpInteraction>;

  startDiff(diffId: string, captureId: string): Promise<StartDiffResult>;
}

////////////////////////////////////////////////////////////////////////////////
export interface IListDiffsResponse {
  diffs: any[];
}

export interface IListUnrecognizedUrlsResponse {
  urls: IUnrecognizedUrl[];
}

export interface IUnrecognizedUrl {
  path: string;
  method: string;
  count: number;
}

export interface IOpticDiffService {
  listDiffs(): Promise<IListDiffsResponse>;

  listUnrecognizedUrls(): Promise<IListUnrecognizedUrlsResponse>;

  learnUndocumentedBodies(
    pathId: string,
    method: string,
    newPathCommands: any[]
  ): Promise<ILearnedBodies>;

  learnShapeDiffAffordances(): Promise<IAffordanceTrailsDiffHashMap>;
}

////////////////////////////////////////////////////////////////////////////////

export interface IOpticDiffRepository {
  findById(id: string): Promise<IOpticDiffService>;

  add(id: string, diff: IOpticDiffService): Promise<void>;
}

////////////////////////////////////////////////////////////////////////////////
export interface IOpticConfigRepository {
  addIgnoreRule(rule: string): Promise<void>;
  getApiName(): Promise<string>;
  listIgnoreRules(): Promise<string[]>;
}

export interface IOpticInteractionsRepository {
  listById(id: string): Promise<any[]>;

  set(id: string, interactions: any[]): Promise<void>;
}

////////////////////////////////////////////////////////////////////////////////

export interface IOpticContext {
  opticEngine: IOpticEngine;
  configRepository: IOpticConfigRepository;
  specRepository: IOpticSpecReadWriteRepository;
  capturesService: IOpticCapturesService;
  diffRepository: IOpticDiffRepository;
}

////////////////////////////////////////////////////////////////////////////////

export interface IBaseSpectacle {
  query<Result, Input = {}>(
    options: SpectacleInput<Input>
  ): Promise<ExecutionResult<Result>>;

  mutate<Result, Input = {}>(
    options: SpectacleInput<Input>
  ): Promise<ExecutionResult<Result>>;
}

export interface IForkableSpectacle extends IBaseSpectacle {
  fork(newSamples?: any[]): Promise<IForkableSpectacle>;
}

// Record<ShapeId, ShapeChoice[]>
export type ShapeViewerProjection = Record<string, ShapeChoice[]>;

export interface SpectacleInput<
  T extends {
    [key: string]: any;
  }
> {
  query: string;
  variables: T;
  operationName?: string;
}

export type GraphQLContext = {
  spectacleContext: () => {
    opticContext: IOpticContext;
    endpointsQueries: endpoints.GraphQueries;
    shapeQueries: shapes.GraphQueries;
    shapeViewerProjection: ShapeViewerProjection;
    contributionsProjection: ContributionsProjection;
    commandGenerator: CommandGenerator;
  };
};
