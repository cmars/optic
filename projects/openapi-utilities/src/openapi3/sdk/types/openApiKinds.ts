export enum OpenApiKind {
  Operation = 'operation',
  Request = 'request',
  QueryParameter = 'query-parameter',
  PathParameter = 'path-parameter',
  HeaderParameter = 'header-parameter',
  ResponseHeader = 'response-header',
  Response = 'response',
  Body = 'body',
  Object = 'object',
  Field = 'field',
  Array = 'array',
  Primitive = 'primitive',
}

export type OpenApiParameterKind = Extract<
  OpenApiKind,
  | OpenApiKind.HeaderParameter
  | OpenApiKind.PathParameter
  | OpenApiKind.QueryParameter
>;

// allow for iterations and `.includes` calls
export const OpenApiParameterKind: OpenApiKind[] = [
  OpenApiKind.HeaderParameter,
  OpenApiKind.PathParameter,
  OpenApiKind.QueryParameter,
];
