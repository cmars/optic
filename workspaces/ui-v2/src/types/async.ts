export type AsyncStatus<T, E = Error> =
  | {
      loading: true;
      error?: undefined;
      data?: undefined;
    }
  | {
      loading: false;
      error: E;
      data?: undefined;
    }
  | { loading: false; error?: undefined; data: T };
