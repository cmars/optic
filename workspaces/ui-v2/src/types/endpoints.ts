import { ChangelogCategory } from '<src>/hooks/useEndpointsChangelog';

export interface IPathParameter {
  id: string;
  name: string;
  isParameterized: boolean;
  description: string;
  endpointId: string;
}

export interface IEndpoint {
  pathId: string;
  method: string;
  purpose: string;
  description: string;
  fullPath: string;
  pathParameters: IPathParameter[];
  group: string;
  isRemoved: boolean;
}

export interface IEndpointWithChanges extends IEndpoint {
  changes: ChangelogCategory | null;
}
