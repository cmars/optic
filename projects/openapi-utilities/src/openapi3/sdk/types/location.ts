import { OpenApiKind } from './openApiKinds';

type ConceptualLocationBase = {
  path: string;
  method: string;
};

export type OperationLocation = ConceptualLocationBase;

export type QueryParameterLocation = ConceptualLocationBase & {
  inRequest: {
    query: string;
  };
};

export type PathParameterLocation = ConceptualLocationBase & {
  inRequest: {
    path: string;
  };
};

export type HeaderParameterLocation = ConceptualLocationBase & {
  inRequest: {
    header: string;
  };
};

export type ResponseHeaderLocation = ConceptualLocationBase & {
  inResponse: {
    header: string;
    statusCode: string;
  };
};

export type ResponseLocation = ConceptualLocationBase & {
  inResponse: {
    statusCode: string;
  };
};

export type BodyLocation = ConceptualLocationBase &
  // Request body
  (| {
        inRequest: {
          body: {
            contentType: string;
          };
        };
      }
    | {
        inResponse: {
          body: {
            contentType: string;
          };
          statusCode: string;
        };
      }
  );

export type FieldLocation = ConceptualLocationBase &
  (
    | {
        inRequest: {
          body: {
            contentType: string;
          };
        };
        jsonSchemaTrail: string[];
      }
    | {
        inResponse: {
          body: {
            contentType: string;
          };
          statusCode: string;
        };
        jsonSchemaTrail: string[];
      }
  );

export type IPathComponent = string;

export type ILocation = {
  jsonPath: string;
  conceptualPath: IPathComponent[];
} & (
  | {
      conceptualLocation: OperationLocation;
      kind: OpenApiKind.Operation;
    }
  | {
      conceptualLocation: QueryParameterLocation;
      kind: OpenApiKind.QueryParameter;
    }
  | {
      conceptualLocation: PathParameterLocation;
      kind: OpenApiKind.PathParameter;
    }
  | {
      conceptualLocation: HeaderParameterLocation;
      kind: OpenApiKind.HeaderParameter;
    }
  | {
      conceptualLocation: ResponseHeaderLocation;
      kind: OpenApiKind.ResponseHeader;
    }
  | { conceptualLocation: ResponseLocation; kind: OpenApiKind.Response }
  | { conceptualLocation: BodyLocation; kind: OpenApiKind.Body }
  | { conceptualLocation: FieldLocation; kind: OpenApiKind.Field }
);
