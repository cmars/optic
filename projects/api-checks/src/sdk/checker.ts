import { DocsLinkHelper, Result, runCheck } from "./types";
import { IChange, OpenApiFact } from "@useoptic/openapi-utilities/build/openapi3/sdk/types";

export class Checker {
  private checkResults: Result[] = [];
  async runCheck(
    change: IChange<OpenApiFact>,
    docsLink: DocsLinkHelper,
    where: string,
    condition: string,
    must: boolean,
    handler: (() => void) | (() => Promise<void>)
  ) {
    this.checkResults.push(
      await runCheck(change, docsLink, where, condition, must, handler)
    );
  }

  listResults() {
    return this.checkResults;
  }
}
