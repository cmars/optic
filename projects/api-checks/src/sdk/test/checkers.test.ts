import { Checker } from "../checker";
import { IChange, OpenApiFact } from "@useoptic/openapi-utilities";
import { newDocsLinkHelper } from "../types";
const { assert } = require("chai"); // Using Assert style

const change: IChange<OpenApiFact> = {
  location: {
    kind: "simulated",
    jsonPath: "/",
    conceptualPath: [],
    conceptualLocation: {
      path: "/simulated",
      method: "get",
    },
  },
} as any;

it("checks can run / fail using off-the-shelf test helpers", async (done) => {
  const check = new Checker();
  const docsHelper = newDocsLinkHelper();
  await check.runCheck(
    change,
    docsHelper,
    "location",
    "is a thing",
    true,
    () => {
      docsHelper.includeDocsLink("https://gitlab.com");
      assert(false, "it's broken!!!");
    }
  );
  const docsHelper2 = newDocsLinkHelper();
  await check.runCheck(
    change,
    docsHelper2,
    "location",
    "is a an ok thing",
    true,
    () => {
      docsHelper2.includeDocsLink("https://github.com");
      assert(true, "it's broken if you see this");
    }
  );

  expect(check.listResults()).toMatchSnapshot();
  done();
});
