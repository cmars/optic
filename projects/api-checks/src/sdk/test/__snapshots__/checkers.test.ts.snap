// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`checks can run / fail using off-the-shelf test helpers 1`] = `
Array [
  Object {
    "change": Object {
      "location": Object {
        "conceptualLocation": Object {
          "method": "get",
          "path": "/simulated",
        },
        "conceptualPath": Array [],
        "jsonPath": "/",
        "kind": "simulated",
      },
    },
    "condition": "is a thing",
    "docsLink": "https://gitlab.com",
    "effectiveOnDate": undefined,
    "error": "it's broken!!!",
    "isMust": true,
    "isShould": false,
    "passed": false,
    "where": "location",
  },
  Object {
    "change": Object {
      "location": Object {
        "conceptualLocation": Object {
          "method": "get",
          "path": "/simulated",
        },
        "conceptualPath": Array [],
        "jsonPath": "/",
        "kind": "simulated",
      },
    },
    "condition": "is a an ok thing",
    "docsLink": "https://github.com",
    "effectiveOnDate": undefined,
    "isMust": true,
    "isShould": false,
    "passed": true,
    "where": "location",
  },
]
`;
